File: leetcode.com/0912_sort_an_array/solution.c
int partition(int *a, int l, int r)
{
    int v  = a[(l+r) / 2];

    while (l <= r) {
        while (a[l] < v)
            l++;
        while (a[r] > v)
            r--;

        if (l >= r)
            break;

        int tmp = a[r];
        a[r] = a[l];
        a[l] = tmp;

        l++;
        r--;
    }

    return r;
}

void sort(int *a, int l, int r)
{
    if (l >= r)
        return;

    int q = partition(a, l, r);
    sort(a, l, q);
    sort(a, q + 1, r);
}

int *sortArray(int *nums, int numsSize, int *returnSize)
{
    sort(nums, 0, numsSize - 1);
    *returnSize = numsSize;
    return nums;
}

File: leetcode.com/0912_sort_an_array/solution.go
func quicksort(nums []int, l, r int) {
    if l >= r {
        return
    }

    val := nums[(l + r) / 2]

    i := l
    j := r

    for i <= j {
        for ; nums[i] < val ; i++ {}
        for ; nums[j] > val ; j-- {}

        if i >= j {
            break
        }

        nums[i], nums[j] = nums[j], nums[i]
        i++
        j--
    }

    quicksort(nums, l, j)
    quicksort(nums, j + 1, r)
}

func sortArray(nums []int) []int {
    quicksort(nums, 0, len(nums) - 1)
    return nums
}
